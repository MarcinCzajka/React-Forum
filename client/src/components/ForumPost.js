import React from 'react';
import basePath from '../api/basePath';
import { Comment, Form, Button } from "semantic-ui-react";
import moment from "moment";
import './ForumPost.css';
import ChildrenOfPost from './ChildrenOfPost';

class ForumPost extends React.Component {
	constructor(props) {
		super(props);
		
		this.state = { 
			id: this.props.postId,
			authorId: "",
			content: "",
			date: "",
			authorNick: "",
			avatar: "https://docs.appthemes.com/files/2011/08/gravatar-grey.jpg",
			cssVisibility: "hidden",
			replyContent: "",
			refreshChildren: false
		};
	}

	refreshChildren() {
		this.setState({
			refreshChildren: !this.state.refreshChildren
		});
	};
	
	componentDidMount() {
		this.getPostDetails();
	}
	
	render() {
		const postsNotToRender = this.props.postsNotToRender.slice();
		let shouldPostRender = postsNotToRender.findIndex(id => {
          return id === this.state.id;
		});
        if (shouldPostRender !== -1) {
            return "";
		}
		
		return (
			<div className="ui large comments">
				<Comment className="comment">
					<Comment.Avatar 
						className="avatar" 
						src={this.state.avatar} >
					</Comment.Avatar>

					<Comment.Content>
						<Comment.Author className="author" as='a'>{this.state.authorNick}</Comment.Author>
						<Comment.Metadata className="metadata">
							<span className="date">{this.state.date}</span>
						</Comment.Metadata>
						<Comment.Text as='p' className="text">{this.state.content}</Comment.Text>

						<Comment.Actions>
							<Button size='mini' onClick={this.changeReplyFormVisibility}>Reply</Button>
							<Button size='mini' onClick={this.removeThisPost}>Delete</Button>
						</Comment.Actions>

						<Form reply className={this.state.cssVisibility}>
							<Form.TextArea value={this.state.replyContent} onChange={e => this.updateReplyTextboxContent(e.target.value)} />
							<Button onClick={this.handleReplyToPost} content='Add Reply' labelPosition='left' icon='edit' primary />
						</Form>
					</Comment.Content>

				</Comment>
				<ChildrenOfPost
					parentId={this.state.id}
					refreshChildren={this.state.refreshChildren}
					handleReplyToPost={this.handleReplyToPost}
					removePostFromState={this.props.removePostFromState}
					addPostToState={this.props.addPostToState}
					postsNotToRender={this.props.postsNotToRender}>
				</ChildrenOfPost>
			</div>
		);
	}

	changeReplyFormVisibility = () => {
		this.setState({cssVisibility: this.state.cssVisibility === "hidden" ? "shown" : "hidden"});
	}

	updateReplyTextboxContent = content => {
		this.setState({replyContent: content});
	};


	handleReplyToPost = async (replyToId, replyMessage) => {
		await basePath({
		  method: "post",
		  url: `/api/posts/`,
		  data: {
			  authorId: "5d1b9e227d1217155c9ba4fe",
			  content: this.state.replyContent,
			  responseTo: this.state.id
		  },
		  withCredentials: true
	  })
	  .then((res) => {
		if(res.status === 200) {
			this.refreshChildren();
			this.changeReplyFormVisibility();
			this.setState({replyContent: ""});
		};
	  });
	};

	getPostDetails = async () => {
		await basePath({
			method: "get",
			url: `/api/posts/${this.state.id}`
		})
		.then(res => {
			this.setState({
				authorId: res.data.authorId || "",
				content: res.data.content || "",
				date: moment(res.data.date)
					.format("MMMM Do YYYY, h:mm:ss")
					.toString()
			});
		})
		.then(() => {
			this.getPostAuthorDetails()
		});
	};

	getPostAuthorDetails = async () => {
		await basePath({
			method: "get",
			url: `/api/users/${this.state.authorId}`
		})
		.then(res => {
			this.setState({
				authorNick: res.data.name || "",
				avatar: res.data.avatar || ""
			});
		})
		.catch(err => {
			this.setState({
				authorNick: "Deleted user."
			});
		});
		this.props.addPostToState(this.state.id);
	};

	removeThisPost = async () => {
		await basePath({
			method: "delete",
			url: `/api/posts/${this.state.id}`,
            withCredentials: true
		})
		.then(res => {
			if(res.status === 200){
				this.props.removePostFromState(this.state.id);
			};
		});
	};

};

export default ForumPost;