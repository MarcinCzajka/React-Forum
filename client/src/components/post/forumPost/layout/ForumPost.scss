:root {
    --roomBorder: none;
    --roomBorder1: 2px solid rgba(6, 12, 15, 0.33);
}

.roomContainer {
    min-width: 60vw;
    max-width: 850px;
    width: 100%;
}

.roomGrid {
    display: grid;
    grid-gap: 0;
    grid-template-columns: repeat(12, 1fr);
    grid-template-rows: repeat(5, 1fr) auto;
}

.roomTitle {
    grid-row: 1;
    grid-column: 5 / -1;

    border-top: var(--roomBorder);
    border-right: var(--roomBorder);
    border-bottom: var(--roomBorder);
}

.roomDescription {
    grid-row: 2 / 6;
    grid-column: 5 / -1;
    border-right: var(--roomBorder);
    border-bottom: var(--roomBorder);
    padding: 10px;
    text-align: left;
}

.roomImageContainer {
    position: relative;
    overflow: hidden;
    width: 100%;
    height: 100%;
    border: var(--roomBorder);

    cursor: pointer;

    grid-row: 1 / 7;
    grid-column: 1 / 5;
}

.roomImage {
    position: absolute;
    top: -9999px;
    bottom: -9999px;
    left: -9999px;
    right: -9999px;
    margin: auto;
    height: 100%;
    min-width: 100%;
    object-fit: cover;
}

.imageOverlay {
    position: relative;
    width: 100%;
    height: 100%;
    opacity: 0;
    background-color: #111;
    
    transition: all 0.3s ease;
}

.imageOverlay > i {
    display: block !important;
    position: absolute;
    top: 50%;
    margin-top: -0.5em !important;
    left: 50%;
    margin-left: -0.5em !important;
}

.imageOverlay:hover {
    opacity: 0.7;
}

.roomFooter {
    grid-row: 6;
    grid-column: 5 / -1;

    border-bottom: var(--roomBorder);
    border-right: var(--roomBorder);
}

.roomStats {
    justify-content: flex-end;
}

.roomStat {
    margin: 3px 10px 3px 10px !important;
}

.forumRoomLink {
    color: initial !important;
}

.roomTextContainer {
    max-height: 200px;
    min-height: 100%;
}

.leftStatistic {
    font-size: 11px;
    margin: 0 auto 0 15px !important;
    align-self: center !important;
}

.leftStatistic > p {
    margin: 0;
}

@media (max-width: 960px) {

    .roomGrid {
        grid-template-columns: 1fr;
        grid-template-rows: repeat(4, auto);
    }

    .roomTitle {
        grid-row: 1;
        grid-column: 1 / -1;

        border-left: var(--roomBorder);
        border-bottom: 0;
    }

    .roomDescription {
        grid-row: 3;
        grid-column: 1 / -1;

        border-left: var(--roomBorder);
    }

    .roomImageContainer {
        width: 100%;
        height: 50vh;
        grid-row: 2;
        grid-column: 1 / -1;
    }

    .roomImage {
        height: 100%;
        min-width: 100%;
        object-fit: contain;
    }

    .roomFooter {
        grid-row: 4;
        grid-column: 1 / -1;

        border-left: var(--roomBorder);
    }

    .roomTextContainer {
        max-height: 200px;
        min-height: 100%;
    }

}