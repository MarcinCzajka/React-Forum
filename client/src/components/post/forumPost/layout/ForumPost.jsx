import React from 'react';
import PropTypes from 'prop-types';
import { Link } from 'react-router-dom';
import { Icon } from "semantic-ui-react";
import ForumPostFooter from './footer/ForumPostFooter';
import ImageModal from '../../../imageModal/ImageModal';
import './ForumPost.scss';

const ForumPost = props => (
	<div className='roomGrid noMargin noPadding'
		style={{visibility: (props.isLoading ? 'hidden' : 'visible'), position: (props.isLoading ? 'absolute' : 'relative')}}
	>

		<div className='roomImageContainer' onClick={props.showImageModal}>
			<img 
				className='roomImage'
				onLoad={props.handleImageLoaded}
				onError={props.removePost}
				src={props.image}
				alt={props.title}
			/>
			<div className='imageOverlay'>
				<Icon name='zoom-in' inverted size='huge' />
			</div>
		</div>

		<header className='roomTitle'>
			<Link to={`/post/${props._id}`}>
				<h4>{props.title}</h4>
			</Link>
		</header>

		<main className='roomDescription'>
			<p>{props.description}</p>
		</main>

		<footer className='roomFooter'>
			<ForumPostFooter 
				_id={props._id}
				authorNick={props.authorNick}
				creationDate={props.creationDate}
				commentsCount={props.commentsCount}
				views={props.views}
				upvotes={props.upvotes}
				isLikedByUser={props.isLikedByUser}
				updateUpvote={props.updateUpvote}
			/>
		</footer>

		<ImageModal image={props.image} alt={props.title} ref={props.imageModalRef} />

	</div>
)

ForumPost.propTypes = {
    _id: PropTypes.string.isRequired,
	image: PropTypes.string.isRequired,
	title: PropTypes.string.isRequired,
	description: PropTypes.string.isRequired,
	showImageModal: PropTypes.func.isRequired,
	removePost: PropTypes.func.isRequired
}

export default ForumPost