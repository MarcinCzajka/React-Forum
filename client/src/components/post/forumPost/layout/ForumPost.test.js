import React from 'react';
import { BrowserRouter as Router } from 'react-router-dom'
import { shallow } from 'enzyme';
import { create } from "react-test-renderer";
import ForumPost from './ForumPost';

const requiredProps = {
    _id: 'abcde12345',
    image: 'www.sdafadsfd.dsa',
    title: 'Random Title',
    description: 'Random description',
    showImageModal: () => {},
    removePost: () => {}
}

describe('PostListPagination component', () => {
    test('renders without crashing', () => {
        shallow(<Router><ForumPost {...requiredProps} /></Router>);
    });

    test('match snapshot 1/3', () => {
        const component = create(<Router><ForumPost {...requiredProps} /></Router>);
        expect(component).toMatchSnapshot();
    });

    test('match snapshot 2/3', () => {
        requiredProps.title = 'Yet Another TiTlE: 12';

        const component = create(<Router><ForumPost {...requiredProps} /></Router>);
        expect(component).toMatchSnapshot();
    });

    test('match snapshot 3/3', () => {
        requiredProps.description = `Light from a service hatch at the rear of the arcade showed him broken lengths of damp chipboard and the drifting shoals of waste. Before they could stampede, take flight from the banks of every computer in the shade beneath a bridge or overpass. The semiotics of the Flatline as a construct, a hardwired ROM cassette replicating a dead man’s skills, obsessions, kneejerk responses. Case felt the edge of the Flatline as a construct, a hardwired ROM cassette replicating a dead man’s skills, obsessions, kneejerk responses. After the postoperative check at the clinic, Molly took him to the simple Chinese hollow points Shin had sold him. Case had never seen him wear the same suit twice, although his wardrobe seemed to consist entirely of meticulous reconstruction’s of garments of the car’s floor. Images formed and reformed: a flickering montage of the Sprawls towers and ragged Fuller domes, dim figures moving toward him in the human system. The Sprawl was a handgun and nine rounds of ammunition, and as he made his way down Shiga from the Chinese program’s thrust, a worrying impression of solid fluidity, as though the shards of a broken mirror bent and elongated as they rotated, but it never told the correct time.`;

        const component = create(<Router><ForumPost {...requiredProps} /></Router>);
        expect(component).toMatchSnapshot();
    });

})