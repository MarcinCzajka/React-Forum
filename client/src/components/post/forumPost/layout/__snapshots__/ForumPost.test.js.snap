// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PostListPagination component match snapshot 1/3 1`] = `
<div
  className="roomGrid noMargin noPadding"
  style={
    Object {
      "position": "relative",
      "visibility": "visible",
    }
  }
>
  <div
    className="roomImageContainer"
    onClick={[Function]}
  >
    <img
      alt="Random Title"
      className="roomImage"
      onError={[Function]}
      src="www.sdafadsfd.dsa"
    />
    <div
      className="imageOverlay"
    >
      <i
        aria-hidden="true"
        className="zoom-in huge inverted icon"
        onClick={[Function]}
      />
    </div>
  </div>
  <header
    className="roomTitle"
  >
    <a
      href="/post/abcde12345"
      onClick={[Function]}
    >
      <h4>
        Random Title
      </h4>
    </a>
  </header>
  <main
    className="roomDescription"
  >
    <p>
      Random description
    </p>
  </main>
  <footer
    className="roomFooter"
  >
    <div
      className="ui mini statistics maxWidth roomStats noMargin"
    >
      <div
        className="ui statistic leftStatistic"
      >
        <p>
          Author: 
          abcde asddf
        </p>
        <p>
          Created: 
          2019-12-17
        </p>
      </div>
      <div
        className="ui statistic roomStat"
      >
        <div
          className="value"
        >
          <a
            href="/post/abcde12345"
            onClick={[Function]}
          >
            <i
              aria-hidden="true"
              className="comments outline icon"
              onClick={[Function]}
            >
               
            </i>
          </a>
        </div>
      </div>
      <div
        className="ui statistic roomStat"
      >
        <div
          className="value"
        >
          <i
            aria-hidden="true"
            className="eye icon"
            onClick={[Function]}
          />
            
          0
        </div>
      </div>
      <div
        className="ui statistic roomStat"
        onClick={[Function]}
        style={
          Object {
            "cursor": "pointer",
          }
        }
      >
        <div
          className="value"
        >
          <i
            aria-hidden="true"
            className="thumbs up icon"
            onClick={[Function]}
            style={
              Object {
                "color": "",
              }
            }
          />
            
          0
        </div>
      </div>
    </div>
  </footer>
</div>
`;

exports[`PostListPagination component match snapshot 2/3 1`] = `
<div
  className="roomGrid noMargin noPadding"
  style={
    Object {
      "position": "relative",
      "visibility": "visible",
    }
  }
>
  <div
    className="roomImageContainer"
    onClick={[Function]}
  >
    <img
      alt="Yet Another TiTlE: 12"
      className="roomImage"
      onError={[Function]}
      src="www.sdafadsfd.dsa"
    />
    <div
      className="imageOverlay"
    >
      <i
        aria-hidden="true"
        className="zoom-in huge inverted icon"
        onClick={[Function]}
      />
    </div>
  </div>
  <header
    className="roomTitle"
  >
    <a
      href="/post/abcde12345"
      onClick={[Function]}
    >
      <h4>
        Yet Another TiTlE: 12
      </h4>
    </a>
  </header>
  <main
    className="roomDescription"
  >
    <p>
      Random description
    </p>
  </main>
  <footer
    className="roomFooter"
  >
    <div
      className="ui mini statistics maxWidth roomStats noMargin"
    >
      <div
        className="ui statistic leftStatistic"
      >
        <p>
          Author: 
          abcde asddf
        </p>
        <p>
          Created: 
          2019-12-17
        </p>
      </div>
      <div
        className="ui statistic roomStat"
      >
        <div
          className="value"
        >
          <a
            href="/post/abcde12345"
            onClick={[Function]}
          >
            <i
              aria-hidden="true"
              className="comments outline icon"
              onClick={[Function]}
            >
               
            </i>
          </a>
        </div>
      </div>
      <div
        className="ui statistic roomStat"
      >
        <div
          className="value"
        >
          <i
            aria-hidden="true"
            className="eye icon"
            onClick={[Function]}
          />
            
          0
        </div>
      </div>
      <div
        className="ui statistic roomStat"
        onClick={[Function]}
        style={
          Object {
            "cursor": "pointer",
          }
        }
      >
        <div
          className="value"
        >
          <i
            aria-hidden="true"
            className="thumbs up icon"
            onClick={[Function]}
            style={
              Object {
                "color": "",
              }
            }
          />
            
          0
        </div>
      </div>
    </div>
  </footer>
</div>
`;

exports[`PostListPagination component match snapshot 3/3 1`] = `
<div
  className="roomGrid noMargin noPadding"
  style={
    Object {
      "position": "relative",
      "visibility": "visible",
    }
  }
>
  <div
    className="roomImageContainer"
    onClick={[Function]}
  >
    <img
      alt="Yet Another TiTlE: 12"
      className="roomImage"
      onError={[Function]}
      src="www.sdafadsfd.dsa"
    />
    <div
      className="imageOverlay"
    >
      <i
        aria-hidden="true"
        className="zoom-in huge inverted icon"
        onClick={[Function]}
      />
    </div>
  </div>
  <header
    className="roomTitle"
  >
    <a
      href="/post/abcde12345"
      onClick={[Function]}
    >
      <h4>
        Yet Another TiTlE: 12
      </h4>
    </a>
  </header>
  <main
    className="roomDescription"
  >
    <p>
      Light from a service hatch at the rear of the arcade showed him broken lengths of damp chipboard and the drifting shoals of waste. Before they could stampede, take flight from the banks of every computer in the shade beneath a bridge or overpass. The semiotics of the Flatline as a construct, a hardwired ROM cassette replicating a dead man’s skills, obsessions, kneejerk responses. Case felt the edge of the Flatline as a construct, a hardwired ROM cassette replicating a dead man’s skills, obsessions, kneejerk responses. After the postoperative check at the clinic, Molly took him to the simple Chinese hollow points Shin had sold him. Case had never seen him wear the same suit twice, although his wardrobe seemed to consist entirely of meticulous reconstruction’s of garments of the car’s floor. Images formed and reformed: a flickering montage of the Sprawls towers and ragged Fuller domes, dim figures moving toward him in the human system. The Sprawl was a handgun and nine rounds of ammunition, and as he made his way down Shiga from the Chinese program’s thrust, a worrying impression of solid fluidity, as though the shards of a broken mirror bent and elongated as they rotated, but it never told the correct time.
    </p>
  </main>
  <footer
    className="roomFooter"
  >
    <div
      className="ui mini statistics maxWidth roomStats noMargin"
    >
      <div
        className="ui statistic leftStatistic"
      >
        <p>
          Author: 
          abcde asddf
        </p>
        <p>
          Created: 
          2019-12-17
        </p>
      </div>
      <div
        className="ui statistic roomStat"
      >
        <div
          className="value"
        >
          <a
            href="/post/abcde12345"
            onClick={[Function]}
          >
            <i
              aria-hidden="true"
              className="comments outline icon"
              onClick={[Function]}
            >
               
            </i>
          </a>
        </div>
      </div>
      <div
        className="ui statistic roomStat"
      >
        <div
          className="value"
        >
          <i
            aria-hidden="true"
            className="eye icon"
            onClick={[Function]}
          />
            
          0
        </div>
      </div>
      <div
        className="ui statistic roomStat"
        onClick={[Function]}
        style={
          Object {
            "cursor": "pointer",
          }
        }
      >
        <div
          className="value"
        >
          <i
            aria-hidden="true"
            className="thumbs up icon"
            onClick={[Function]}
            style={
              Object {
                "color": "",
              }
            }
          />
            
          0
        </div>
      </div>
    </div>
  </footer>
</div>
`;
